src/output-writer/index.ts

```diff
@@ -1,13 +1,154 @@
+    };
+
+export interface Column {
+  name: string;
+  /** Extra details to go along with the column name, usually the unit */
+  nameDetail?: string;
+  lighthouseCategory: string;
+  field: ColumnField;
+}
 
 export interface OutputWriter {
-  addEntry(report: LHR): Promise<void>;
+  writeRunInfo?: (runInfo: RunInfo) => Promise<void>;
+  writeHeader(columns: Column[]): Promise<void>;
+  addEntry(url: string, rowValues: string[]): Promise<void>;
   complete(): Promise<void>;
 }
 
+/**
+ * Wraps an OutputWriter to allow being passed LHR (lighthouse reports)
+ * and adapts it into a row/column format that is shared
+ * between the different output writers.
+ * Also, it handles calling writeHeader and writeRunInfo automatically after the first addEntry call.
+ */
+export const adaptLHRToOutputWriter = (
+  outputWriter: OutputWriter,
+  command: string,
+  lighthouseParadeVersion: string
+) => {
+  // Used to make sure that the addEntry calls happen one at a time
+  // (specifically important to make sure the header finishes getting written
+  // before the next row starts getting written)
+  let mutexPromise: Promise<unknown> = Promise.resolve();
+  let hasWrittenHeader = false;
+  const columns: Column[] = [];
+
+  return {
+    complete: () => outputWriter.complete(),
+    async addEntry(report: LHR) {
+      if (!hasWrittenHeader) {
+        mutexPromise = mutexPromise.then(async () => {
+          hasWrittenHeader = true;
+          for (const category of Object.values(report.categories)) {
+            columns.push({
+              name: category.title,
+              nameDetail: 'category score',
+              lighthouseCategory: category.title,
+              field: {
+                type: ColumnType.CategoryScore,
+                category: category.id,
+              },
+            });
+          }
+
+          for (const category of Object.values(report.categories)) {
+            for (const audit of Object.values(category.auditRefs)) {
+              const auditData = report.audits[audit.id];
+              if (auditData.scoreDisplayMode === 'numeric') {
+                columns.push({
+                  name: auditData.title,
+                  nameDetail: 'score',
+                  lighthouseCategory: category.title,
+                  field: {
+                    type: ColumnType.AuditScore,
+                    hasAuditValueColumn: auditData.numericValue !== undefined,
+                    audit: audit.id,
+                  },
+                });
+              }
+
+              if (auditData.numericValue) {
+                columns.push({
+                  name: auditData.title,
+                  nameDetail: auditData.numericUnit,
+                  lighthouseCategory: category.title,
+                  field: {
+                    type: ColumnType.AuditValue,
+                    unit: auditData.numericUnit || '',
+                    audit: audit.id,
+                  },
+                });
+              }
```

Comment:
Yes there may be 0, 1, or 2 columns. Most of them will have 2 (one that is the score and one that is the numeric value).

Reply:
Gotcha, okay that makes sense. It might be nice to add a little more context in a comment since it's not obvious if you're not already familiar with the type of data in the report.